xbmc/addons/BinaryAddonCache.cpp

```diff
@@ -0,0 +1,78 @@
+/*
+ *      Copyright (C) 2016 Team Kodi
+ *      http://kodi.tv
+ *
+ *  This Program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2, or (at your option)
+ *  any later version.
+ *
+ *  This Program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with XBMC; see the file COPYING.  If not, see
+ *  <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "BinaryAddonCache.h"
+#include "AddonManager.h"
+#include "threads/SingleLock.h"
+
+namespace ADDON
+{
+
+CBinaryAddonCache::~CBinaryAddonCache()
+{
+  CAddonMgr::GetInstance().UnregisterObserver(this);
+}
+
+void CBinaryAddonCache::Init()
+{
+  Update();
+  CAddonMgr::GetInstance().RegisterObserver(this);
+}
+
+void CBinaryAddonCache::GetAddons(VECADDONS& addons, const TYPE& type)
+{
+  CSingleLock lock(m_critSection);
+  auto it = m_addons.find(type);
+
+  if (it != m_addons.end())
+  {
+    for (auto &addon : it->second)
+    {
+      addons.push_back(addon);
+    }
+  }
+}
+
+void CBinaryAddonCache::Notify(const Observable &obs, const ObservableMessage msg)
+{
+  Update();
+}
+
+void CBinaryAddonCache::Update()
+{
+  using AddonMap = std::multimap<TYPE, VECADDONS>;
+  AddonMap addonmap;
+  addonmap.clear();
+
+  VECADDONS addons;
+  CAddonMgr::GetInstance().GetAddons(addons, ADDON_AUDIODECODER);
```

Comment:
Maybe initialise a list of addon types to be cached so that you can loop over that list. That makes it easier to add additional addon types and reduces code duplication.